[Spark] Base merge schema evolution off set expressions, not source schema

Kimahriman opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-io/delta/blob/master/CONTRIBUTING.md
  2. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If possible, provide a concise example to reproduce the issue for a faster review.
  6. If applicable, include the corresponding issue number in the PR title and link it in the body.
-->

#### Which Delta project/connector is this regarding?
<!--
Please add the component selected below to the beginning of the pull request title
For example: [Spark] Title of my pull request
-->

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

<!--
- Describe what this PR changes.
- Describe why we need the change.
 
If this PR resolves an issue be sure to include "Resolves #XXX" to correctly link and close the issue upon merge.
-->
Resolves #2300 

This updates schema evolution in merging to use the target assignment columns and expression schema to determine the new target schema instead of the source schema. The current implementation seems to only support upserting use cases, where the source schema looks like the target schema. However when using custom set expressions, there's no requirement for this to be the case. 

When auto schema migration is enabled and the target assignment column doesn't exist in the table, the column parts and the data type of the expression are used for determining schema updates, regardless of what the source schema is. For upserting cases, this works the same way, since the data types and assignment columns will match the source table. But it also supports more use cases where the source schema doesn't match the target schema.

## How was this patch tested?

<!--
If tests were added, say they were added here. Please make sure to test the changes thoroughly including negative and positive cases if possible.
If the changes were tested in any way other than unit tests, please clarify how you tested step by step (ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future).
If the changes were not tested, please explain why.
-->
Updated two tests that were showing this use case didn't work, to show it now does work, and added one additional test for adding a new nested column in an update.

## Does this PR introduce _any_ user-facing changes?

<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Delta Lake versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes, supports more schema evolution use cases in merge

Kimahriman committed:
Use update expressions instead of source schema for merge schema evolution

Kimahriman committed:
scalastyle fix

Kimahriman committed:
Remove non-nullability

Kimahriman commented:
cc @johanl-db since you added the non-star schema evolution in https://github.com/delta-io/delta/commit/affd5774def34bc4d553aff805cd17afe7d57ea1

johanl-db commented:
I actually considered this when implementing non-star schema evolution, the main concern was that schema evolution then becomes very error-prone as any typo will happily add a new column to the target:
```
MERGE INTO target
USING source
WHEN MATCHED THEN UPDATE SET userid = source.user_id
```
where the target already contains a column `user_id` which was misspelled as `userid`. This concern is actually the main reason schema evolution initially only applied to UPDATE SET * / INSERT * and not non-star action.

We could allow it but in my opinion (and others that I discussed this with) is that this would do more harm than good.

The following should enable the use case you describe on the other hand (using the example from https://github.com/delta-io/delta/issues/2300):

```
  spark.createDataFrame([["a"], ["b"], ["c"]], ["old"]).write.format("delta").mode("overwrite").save("merge-test")

  val df = spark.createDataFrame([["a", "1"], ["b", "2"], ["c", "3"]], ["first", "second"])
  val source = df.withColumnRenamed("second", "new")

  val dt = DeltaTable.forPath(spark, "file:/path/to/table/merge-test")
  (dt.alias("target").merge(df.alias("source"), "target.old = source.first")
      .whenMatchedUpdate(set = {
          "new": "source.new"
      })
      .execute()
```

Renaming the source column before merging has the advantage that the intent is explicit:
```
  val source = df.withColumnRenamed("second", "new")
```

Kimahriman commented:
> I actually considered this when implementing non-star schema evolution, the main concern was that schema evolution then becomes very error-prone as any typo will happily add a new column to the target:
> 
> ```
> MERGE INTO target
> USING source
> WHEN MATCHED THEN UPDATE SET userid = source.user_id
> ```
> 
> where the target already contains a column `user_id` which was misspelled as `userid`. This concern is actually the main reason schema evolution initially only applied to UPDATE SET * / INSERT * and not non-star action.

Aren't all operations prone to typos? You could have a DataFrame with `userid` in it and do a `UPDATE SET *` and end up in the same scenario, not sure how this is any different. If someone is that worried about typos, they should be testing their queries or not enable automatic schema evolution (which is disabled by default so this is all already feature-gated).

> The following should enable the use case you describe on the other hand (using the example from #2300):

That was a trivial example. In reality what we are doing is
```python
set={'new.nested.field': complexExpressionInvolvingSourceAndTarget}
```

Getting the source to "look like" the target is non-trivial for our case without doing a pre join, which then just adds another whole join (really two since merge is two passes). And it provides no real benefit. 

Additionally the current implementation has a fun side effect if you deviate from the "copy the exact field" (which is only one use case for merge)

```python
(spark.createDataFrame([["a"], ["b"], ["c"]], ["old"])
    .repartition(1)
    .write
    .format("delta")
    .mode("overwrite")
    .save("merge-test")
)

df = spark.createDataFrame([["a", 1], ["b", 2], ["c", 3]], ["first", "second"]).repartition(1)

dt = DeltaTable.forPath(spark, "file:/path/to/merge-test")
(dt.alias("target").merge(df.alias("source"), "target.old = source.first")
    .whenMatchedUpdate(set = {
        "first": "source.second"
    })   
    .execute()
)

spark.read.format("delta").load("merge-test").printSchema()
spark.read.format("delta").load("merge-test").show()
```

The output:
```
root
 |-- old: string (nullable = true)
 |-- first: string (nullable = true)

+---+-----+
|old|first|
+---+-----+
|  c|    3|
|  b|    2|
|  a|    1|
+---+-----+
```

Even though the target value I inserted was an integer (`second`), `first` is used as the data type to use for schema evolution, adding a string column, and then casting the integer I am inserting to a string. This seems much more problematic than worrying about typos.

johanl-db commented:
Schema evolution allows adding columns/fields that are present in the source dataframe but missing in the table schema, the use case you are describing is different though since the field `new.nested.field` isn't present in the source in the first place.
You would need to explicitly add the column to the table beforehand:
```
DeltaTable.forPath(spark, "file:/path/to/merge-test")
  .addColumn('new.nested.field', dataType= <type>)
  .execute()
```

Relying on the merge clauses to define what the new schema instead of just the target/source schemas comes with some challenges:
- It's a departure from the current definition of schema evolution, likely a breaking change.
- The resulting schema is ambiguous: multiple clauses may introduce the same new column with different types,

Kimahriman commented:
> Schema evolution allows adding columns/fields that are present in the source dataframe but missing in the table schema, the use case you are describing is different though since the field `new.nested.field` isn't present in the source in the first place. You would need to explicitly add the column to the table beforehand:

Yes that is the exact scenario I'm trying to fix since we have all the information we need to do the schema evolution automatically without manually adding the column beforehand. Again this assumes that upserting is the only use case for merge, and that is not the case at all. If that was the case then custom set expressions shouldn't be supported. The source table doesn't have to have anything to do with the target table schema wise.

> 
> Relying on the merge clauses to define what the new schema instead of just the target/source schemas comes with some challenges:
> 
> * It's a departure from the current definition of schema evolution, likely a breaking change.
> * The resulting schema is ambiguous: multiple clauses may introduce the same new column with different types,

The code is simplified and all the existing tests pass, do you have an example you think will cause problems? The new behavior should be a superset of the existing behavior (beside from fixing the example buggy behavior mentioned in the last post). If you update/insert a field as is then it will use the schema as is as well.

johanl-db commented:
User-facing change 1: Assigning to a column that isn't in the target or source now adds that column to the target table instead of failing:
```
target: key int
source: key int
```
```
MERGE INTO target
USING source
ON source.key = target.key
WHEN MATCHED THEN UPDATE SET new_column = source.key
```

User-facing change 2: Assigning to a column that exists in the source but not the target can now result in the column being added with a different type:
```
target: key int
source: key int, value string
```
```
MERGE INTO target
USING source
ON source.key = target.key
WHEN MATCHED THEN UPDATE SET value = source.value::int
```
Currently `value` is added to the target with type `string`, with your proposal it's added with type `int`.

I'm not too concerned about the first case as we're just making schema evolution more permissive (hopefully users don't rely on an error being thrown there). The second one could be breaking some use cases though.

The bigger question remaining is what should the resulting type be if there are multiple clauses introducing a new column?

If one is strictly larger than the other, then arguably we should pick the larger type and upcast the smaller one:
```
WHEN MATCHED AND condition THEN UPDATE SET value = source.value::int
WHEN MATCHED THEN UPDATE SET value = source.value::long
```
`value` should be added with type `long`

What if both types can be upcast to each others:
```
WHEN MATCHED AND condition THEN UPDATE SET value = source.value::timestamp
WHEN MATCHED THEN UPDATE SET value = source.value::timestampNTZ
```
Both `timestamp` and `timestampNTZ` are valid types for the new column `value`, how do we pick one? We could for example catch such a case and throw an error because the resulting type is ambiguous.

Similar with `string` which supports implicit cast to and from most other types:
```
WHEN MATCHED AND condition THEN UPDATE SET value = source.value::string
WHEN MATCHED THEN UPDATE SET value = source.value::date
```
Should `value` in the target table be a `date` or `string`?

If we can get a good answer for type reconciliation and get some confidence that the second user-facing change isn't going to break user workloads then that would clear the path forward for this change

Kimahriman commented:
> User-facing change 1: Assigning to a column that isn't in the target or source now adds that column to the target table instead of failing:

I wouldn't consider this a breaking change, just a new feature. Something that wasn't possible before is now possible.

> User-facing change 2: Assigning to a column that exists in the source but not the target can now result in the column being added with a different type:

Technically a breaking change, but I would consider this a bug fix. I can't imagine a reasonable person would expect this behavior, let alone rely on it.

> The bigger question remaining is what should the resulting type be if there are multiple clauses introducing a new column?
> ...
> If we can get a good answer for type reconciliation and get some confidence that the second user-facing change isn't going to break user workloads then that would clear the path forward for this change

I will add some tests case for this. Because it all uses the `mergeSchemas` helper with `allowImplicitConversions=true`, I would expect it would do type reconciliation to the smallest common type, but the tests cases will find out for sure.

Kimahriman committed:
Add merge evolution reconciliation tests

Kimahriman commented:
So it looks like the current behavior is all based around Spark's `implicitCast`, which itself depends on the order of the types you're merging. Giving this, it seems like there's 4 possible options for the behavior of multiple expressions to the same new field:
1. Use the `implicitCast` behavior, ensuring to process all clauses in order, regardless of update or insert. This behavior is highly dependent on the two types for how it behaves, but it's what the current process uses for adding any new fields into the target table.
2. Use a method like you described of "upcasting" to the common type. I would need to see if there's any existing Delta or Spark helper to do something like this. This would be a little different than the implicit cast. For example merging an int and a long (in that order), would result in an int in the implicitCast method (casting the long to an int), but a long in an "upcasting" approach.
3. Use the first type seen for a given target column and then rely on the casting of any additional uses of that target column.
4. Check that all assignments to the same column are the same data type, and fail if not. Maybe the least ambiguous approach from a user perspective.

I added some unit tests showing the current behavior, which is mostly 1) but without the full ordering across both insert and update

johanl-db commented:
The order of the clauses shouldn't have an impact on the resulting type of a column, that would be arbitrary. This rules out 1. and 3. and leaves 2., the more permissive approach and 4. the stricter one.

The issue with 4. and failing when the clauses use different types is that MERGE operations may start to fail:
```
source: float_col: float, new_column: double
WHEN MATCHED THEN UPDATE SET *
WHEN NOT MATCHED THEN INSERT (new_column) VALUES (source.float_col)
```
This is valid today and will insert `new_column` as `double` but would fail with approach 4.
It also leads to cases that have limitations that seem very arbitrary:
```
WHEN MATCHED AND condition1 THEN UPDATE SET new_column = 5 -- int
WHEN MATCHED AND condition2 THEN UPDATE SET new_column = 10 / 2 -- double
```
fails because the two types are different.

I'd be in favor of the 3rd option, reconciling the different types found but being very careful about what is allowed.
Spark's [AnsiTypeCoercion.findTightestCommonType](https://github.com/apache/spark/blob/2423a4517ace1d76938ee3fd285594900f272552/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala#L97) seems suited, it'll allow common type widening without being too permissive - e.p. it forbids a whole class of ambiguous cases I mentioned in my previous comment which implicit cast or even upcast would allow.

Few things that I see that would need to be covered:
- If the column already exists in the target, then the target type should not be changed.
- Sufficient testing of the type reconciliation logic.
- Struct evolution must work correctly, e.p. adding fields to structs nested in maps/arrays. I don't think we necessarily need special handling in this change but this is an area that was missed in the past and we don't have tests yet that cover adding fields that exists already in the source.

Kimahriman committed:
Use type coersion instead of implicit casting

Kimahriman commented:
Updated to use [TypeCoercion.findWiderTypeForTwo](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L945)

AnsiTypeCoercion will resolve a string and a integral to a long, and not a string. The non-ansi TypeCoercion seemed like a better fit.

I updated the tests I made to suit. Trying to figure out what nested tests to add. For the most part nested things shouldn't be any different.

Kimahriman committed:
Fix spelling of coercion

Kimahriman committed:
Add some complex type tests

Kimahriman committed:
Properly handle nullability merging for type coercion

Kimahriman committed:
Add another array and map test

Kimahriman commented:
Added more complex type tests, let me know what how it looks

Kimahriman committed:
Scalastyle

Kimahriman committed:
Ensure existing fields won't fail due to type coercion

Kimahriman commented on spark/src/main/scala/org/apache/spark/sql/delta/schema/SchemaMergingUtils.scala:
```diff
@@ -157,6 +157,10 @@ object SchemaMergingUtils {
    *                                 the same way in Parquet files. With this flag enabled, the
    *                                 merge will succeed, because once we get to write time Spark SQL
    *                                 will support implicitly converting the int to a string.
+   * @param allowTypeCoercion Whether to allow type coersion using Spark's AnsiTypeCoersion when
```
Thanks, fixed

Kimahriman commented on spark/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/deltaMerge.scala:
```diff
@@ -586,62 +580,39 @@ object DeltaMergeInto {
-      // into a LONG target.
+          // Combine schemas for each action, allowing type coersion to a wider type if the types
+          // differ.
+          SchemaMergingUtils.mergeSchemas(
```
Updated to ensure existing fields don't get coerced

Kimahriman commented on spark/src/main/scala/org/apache/spark/sql/delta/schema/SchemaMergingUtils.scala:
```diff
@@ -166,6 +170,7 @@ object SchemaMergingUtils {
       tableSchema: StructType,
       dataSchema: StructType,
       allowImplicitConversions: Boolean = false,
+      allowTypeCoercion: Boolean = false,
```
Moved

Kimahriman commented on spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoSchemaEvolutionSuite.scala:
```diff
@@ -951,6 +953,103 @@ trait MergeIntoSchemaEvolutionBaseTests {
+    expected = Seq((1, "2")).toDF("key", "value").select('key, struct("value").alias("nested")),
+    expectErrorWithoutEvolutionContains = "cannot resolve nested in UPDATE clause"
+  )
+  // scalastyle:on line.size.limit
```
I added a test for the case you mentioned below of an existing string field that we try to write both a string and bool too

foss-contributor commented:
@Kimahriman, thank you for your contribution! This pull request has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this pull request or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!
